---
title: "chm13 cut&run"
author: "G$"
date: "2/24/2021"
output: html_document
---


```{r include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning=FALSE)
source("~/T2T-Epigenetics/utils/ilee_plot_utils.R")
source("~/T2T-Epigenetics/utils/methylation_R_utils.R")
library(tidyverse)
library(cowplot)
library(BSgenome.t2t.v1.0.release)
list=seqnames(BSgenome.t2t.v1.0.release)
# don't keep chrM, problems arise later because not all beds have chrM
list=list[1:23]
```

```{r mbed}
figs="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/figures"
dat="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly"


censat = read_tsv(paste0(dat, "/annotations/t2t_cenAnnotation.v2.021621.bed"), col_names = F) %>%
  mutate(name = ifelse(grepl("hsat1", X4), "HSAT1", X4)) %>%
  mutate(name = ifelse(grepl("hsat2", X4), "HSAT2", name)) %>%
  mutate(name = ifelse(grepl("hsat3", X4), "HSAT3", name)) %>%
  mutate(name = ifelse(grepl("hsat4", X4), "HSAT4", name)) %>%
  mutate(name = ifelse(grepl("hsat5", X4), "HSAT5", name)) %>%
  mutate(name = ifelse(grepl("hsat6", X4), "HSAT6", name)) %>%
  mutate(name = ifelse(grepl("ct", X4), "CT", name)) %>%
  mutate(name = ifelse(grepl("bsat", X4), "BSAT", name)) %>%
  mutate(name = ifelse(grepl("dhor", X4), "DHOR", name)) %>%
  mutate(name = ifelse(grepl("hor", name), "HOR", name)) %>%
  mutate(name = ifelse(grepl("mon", X4), "MON", name)) %>%
  mutate(name = ifelse(grepl("GSAT", X4), "GSAT", name)) %>%
  dplyr::select(c(X1, X2, X3, name)) %>%
  dplyr::rename("chrom" =1, "start" = 2 ,"end" =3) %>%
  mutate(ID=row_number())

```

```{r}

name="_cutnrun-202021_losalt.F3852.over.IlluminaPCRfree_v1.0-assembly_51mers_single_mrg_meryl_sort.bigwig"
path="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/cut_and_run/210204_alignments/022021_CUTRUN"

bw <- paste0(path,"/H3K9me3/CHM13_H3K9me3",name)
rep1 <- import(bw, format="bigwig")

bw <- paste0(path,"/IgG/CHM13_IgG",name)
igg1 <- import(bw, format="bigwig")

name="_cutnrun_losalt.F3852.over.IlluminaPCRfree_v1.0-assembly_51mers_single_mrg_meryl_sort.bigwig"
path="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/cut_and_run/210204_alignments/122920_CUTRUN"

bw <- paste0(path,"/H3K9me3/CHM13_H3K9me3",name)
rep2 <- import(bw, format="bigwig")

bw <- paste0(path,"/IgG/CHM13_IgG",name)
igg2 <- import(bw, format="bigwig")
```

```{r}
censat.gr <- GRanges(censat)

rep1.censat <- FindOvls(rep1,censat.gr) %>%
  mutate(treat="h3k9me3") %>%
  mutate(rep="rep022021")

igg1.censat <- FindOvls(igg1,censat.gr) %>%
  mutate(treat="igg") %>%
  mutate(rep="rep022021")


# rep 2

rep2.censat <- FindOvls(rep2,censat.gr) %>%
  mutate(treat="h3k9me3") %>%
  mutate(rep="rep122920")

igg2.censat <- FindOvls(igg2,censat.gr) %>%
  mutate(treat="igg") %>%
  mutate(rep="rep122920")

all.h3k9 <- rbind(rep1.censat, rep2.censat, igg1.censat,igg2.censat)
```


```{r}

SAT = c("HSAT1", "HSAT2", "HSAT3","HOR", "CT")

merged_treat <- all.h3k9 %>%
  filter(name %in% SAT)

treat.total1 <- sum(rep1$score)
control.total1 <- sum(igg1$score)
ratio1=control.total1/treat.total1

treat.total2 <- sum(rep2$score)
control.total2 <- sum(igg2$score)
ratio2=control.total2/treat.total2

stat.sum <- merged_treat %>%
  group_by(name, treat,rep) %>%
  summarise(reads=sum(score)) %>%
  spread(treat, reads) %>%
  ungroup() %>%
  group_by(name,rep) %>%
  summarise(enrich=if_else(rep=="rep022021",log2((h3k9me3/igg)*ratio1),log2((h3k9me3/igg)*ratio2)))

write.table(stat.sum, "/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/cut_and_run/H3K9me3_51merintersections_log2FC", row.names = F, quote=F, sep="\t")

ggplot(stat.sum, aes(x=name, y=enrich, color=name))+geom_point(position = "dodge",size=2.5)+ scale_color_manual(values = censatColors, drop = FALSE)+geom_hline(yintercept = 0, linetype="dashed")+facet_wrap(~rep)


ggsave(
  paste0(figs, "/", "CHm13H3K9Me3_enrichment.pdf"),
  plot = last_plot(),
  scale = 1,
  width = 8,
  height = 5,
)

```




