---
title: "load_methyl_bed"
author: "G$"
date: "9/3/2020"
output: pdf_document
---

---
```{r include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning=FALSE)
library(knitr)
library(tidyverse)
library(bsseq)
library(Biostrings)
library(ggplot2)
library(png)
library(cowplot)
options(scipen=999)
library(zoo)
options(knitr.duplicate.label = 'allow')
source("/home/isac/Code/ilee/plot/ilee_plot_utils.R")
library("ggsci")
source("/home/isac/Code/nanopore-methylation-utilities/methylation_R_utils.R")
library(mclust)
```

```{r mbed}
figs="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/figures"
dat="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/HG002"
```

```{r, function}

  # get boundaries of reads if not provided
#bounds <- all_runs %>% group_by(qname) %>%
#      summarize(start = min(start),
#                end = max(end), 
#                freq = freq) %>%
#  distinct()
#  
#
#bounds<- bounds %>% 
#  arrange(freq, start, end) 
#
#    bounds <- bounds %>%
#      ungroup() %>%
#        mutate(
#          readi = as.numeric(seq_len(length(unique(all_runs$qname)))),
#          ymin = -readi - 0.8, 
#          ymax = ymin + 0.6)
#
#  all_runs <- all_runs %>%
#    mutate(ymin = bounds$ymin[match(qname,bounds$qname)],
#           ymax = bounds$ymax[match(qname,bounds$qname)])
#  bounds <- bounds %>%
#    mutate(ymin = bounds$ymin[match(qname,bounds$qname)],
#           ymax = bounds$ymax[match(qname,bounds$qname)])
#  
  
order_reads_facet <- function(x,offset = 0, bounds=NULL, qorder = NULL){
  # get boundaries of reads if not provided
  if (is.null(bounds)){
    bounds <- x %>% group_by(qname, group) %>%
      summarize(start = min(start),
                end = max(end))
    # label y based on order of smallest start
    # if q order is given, use this to order
    if (!is.null(qorder)) {
      bounds <- bounds %>%
        mutate(qname = factor(qname,levels = qorder)) %>%
        arrange(qname) 
    } else {
      bounds<- bounds %>% 
        arrange(start, end) 
    }
    bounds <- bounds %>%
      ungroup() %>%
      group_by(group) %>%
        mutate(
          readi = row_number(),
          ymin = -readi - 0.8 - offset, 
          ymax = ymin + 0.6)
  }
  
  x <- x %>%
    mutate(ymin = bounds$ymin[match(qname,bounds$qname)],
           ymax = bounds$ymax[match(qname,bounds$qname)])
  bounds <- bounds %>%
    mutate(ymin = bounds$ymin[match(qname,bounds$qname)],
           ymax = bounds$ymax[match(qname,bounds$qname)])
  return(list(x = x,bounds = bounds))
}

```


```{r bis, echo=F}
cgcalls <- read_csv(paste0(dat, "/nanonome/methylation_calls/pooled/HG002_nanonome_chrX_CpGmethylation_ByCall.csv"))

chr="chrX"
start=112297349
end=112514845
#size_sel <- reads %>%
#  dplyr::mutate(rlen = end-start) %>%
#  filter(rlen >= 50000) %>%
#  mutate(motif = "CG")

bis <- cgcalls %>% 
  group_by(chrom, start) %>%
  summarise(meth_freq = mean(mcall), cov = n()) %>%
  mutate(smooth = rollmean(meth_freq, 50, fill = NA)) %>%
  mutate(cov_smooth = rollmean(cov, 50, fill = NA))

meth <- ggplot(bis, aes(x = (start/1e6), y= smooth))+geom_line(size =1) + labs(x="Genomic coordinates (Mb)", y="Methylation")+theme_classic(base_size = 25)+ylim(0,1)+xlim((start/1e6),(end/1e6))

cov <- ggplot(bis, aes(x = (start/1e6), y= cov_smooth))+geom_line(size =1) + labs(x="Genomic coordinates (Mb)", y="Coverage")+theme_classic(base_size = 25)+xlim(start,end)

meth


ggsave(
  paste0(figs, "/DXZ4_methyl_patternHG002.pdf"),
  plot = meth,
  scale = 1,
  width = 12,
  height = 6,
)

```


```{r plot, echo=F}


df_meth <- df %>%
  distinct() %>%
  select(c(qname, group)) %>%
  distinct() %>%
  group_by(qname) %>%
  dplyr::slice(1L)

cgcalls_dx <- cgcalls %>%
  filter(start >112350000) %>%
  filter(end < 112380000)

cpg_runs <-getRuns(cgcalls_dx, maxGap = 100)

#all_runs <- merge(cpg_runs,df_meth, by = "qname" )

cpg_runs.ordered <- order_reads(cpg_runs)



cpg_runs_plot <- cpg_runs.ordered$x %>%
    mutate(m = ifelse(values == 1, "Methylated","Unmethylated")) %>%
    mutate(mod = "CpG")
  

pal <- pal_npg("nrc")(10)
meth_pal <- c(pal[8],pal[4]) 
#ids <- order(sapply(seq_along(gcruns.list),function(i){length(unique(gcruns.list$qname))}))


g <- ggplot(cpg_runs_plot,aes(xmin = start, xmax = end, ymin = ymin, ymax = ymax))+ geom_rect(data = cpg_runs.ordered$bounds, fill = "grey80")+ 
    geom_rect(aes(fill = m))+scale_fill_manual(name = "State", values = meth_pal)#+facet_wrap(~group, ncol =1)#+coord_cartesian(xlim=c(114176505, 114211509))+ theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), panel.spacing = unit(2, "lines"))+theme_classic()
  print(g)
  
  ggsave(
paste0(figs, "/DXZ4_phasedReadLevelHG002.pdf"),
        plot = g,
        scale = 1,
        width = 12,
        height = 6,
        )
  
```
