---
title: "load_methyl_bed"
author: "G$"
date: "9/3/2020"
output: pdf_document
---

---
```{r include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning=FALSE)
# loads libs 

source("~/T2T-Epigenetics/utils/ilee_plot_utils.R")
source("~/T2T-Epigenetics/utils/methylation_R_utils.R")
library(tidyverse)
library(cowplot)
library(BSgenome.t2t.v1.0.release)
library(GenomicRanges)
library(knitr)
library(Biostrings)
library(cowplot)
options(scipen=999)
library(zoo)
library(mclust)
```

```{r mbed}
figs="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/revision_analysis/promoter_clustering/figures"
dat="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly"

```

```{r bis, echo=F}

protein_coding <- read_tsv(paste0(dat,"/revision_analysis/promoter_clustering/mitchell_merged_bed/protein_coding_names.txt"),col_names="gene")

reads <- tabix_mbed(paste0(dat, "/revision_analysis/promoter_clustering/chrX/chrX.chm13_v1_tss.CGI_intersectionsMethylBed.bed"),extcol = "motif",by = "read") 

regs.all <- read_delim(paste0(dat, "/revision_analysis/promoter_clustering/mitchell_merged_bed/chm13_v1_tss.CGI_ntranscripts_intersections.bed"), col_names=c("chr","start","end","strand","name","gene_start","gene_end", "transcripts", "cgi_chr","cgi_start","cgi_end"),delim="\t")  %>%
  filter(name %in% protein_coding$gene) %>%
  mutate(quantile=case_when(transcripts == 0 ~ "no_expression", 
                            transcripts > 0 & transcripts < 113 ~ "medium",
                            transcripts >=100 ~ "high")) %>%
  filter(chr=="chrX") 

regs.cgi <- regs.all  %>%
  mutate(region="cgi") 

regs.gb <- regs.all  %>%
  mutate(start=ifelse(strand=="+", cgi_end,gene_end))%>%
  mutate(end=ifelse(strand=="+", gene_end,cgi_end)) %>%
  mutate(len=end-start) %>%
  filter(chr=="chrX") %>%
  filter(len > 100) %>%
  dplyr::select(-c(len))  %>%
  mutate(region="gb")


regs <- rbind(regs.gb,regs.cgi) %>%
  distinct() %>%
 # mutate(start=ifelse(strand=="+", gene_start,gene_end))%>%
#  mutate(end=ifelse(strand=="+", gene_end,gene_start)) %>%
  GRanges()


size_sel <- reads %>%
  dplyr::mutate(rlen = end-start) %>%
  filter(rlen >= 5000) %>%
  mutate(motif = "CG")

cgcalls <- mbedByCall(size_sel) %>%
  drop_na(mcall) 
cgcalls <- as.data.frame(cgcalls) %>%
  mutate(cg_position=start) %>%
  GRanges()

ovls <- FindOvls(regs,cgcalls) 

combined <- ovls %>%
  #group_by(qname) %>%
 # mutate(read_start=min(cg_position), read_end=max(cg_position)) %>%
  #filter(if (region=="cgi") read_start - start < 500 else read_start ==read_start )
  #filter(read_start - start < 500) %>%
  #filter(end - read_end < 500) %>%
  group_by(region,name,qname,transcripts,quantile) %>%
  summarise(methylation=mean(mcall)) %>%
  spread(region,methylation) %>%
  na.omit()

write.table(combined, "/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/revision_analysis/promoter_clustering/chrX/CGI_GeneBody_SingleRead_methylation.tsv",
          quote=F,
          row.names=F,
          sep="\t")


```



```{r}
combined <- read_tsv("/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/revision_analysis/promoter_clustering/chrX/CGI_GeneBody_SingleRead_methylation.tsv")
df.cgi <- combined %>%
  dplyr::select(c(name,qname,cgi)) %>%
  unite(newname,name,qname, sep=":") 

class <- df.cgi$newname
X <- df.cgi$cgi
model1 <- Mclust(X,G=2)
summary(model1)
#plot(model1, what = c("classification"))
df.cgi$cgi_cluster <- model1$classification
df.cgi$uncertainty_cgi <- model1$uncertainty


clustered_reads <- df.cgi%>%
  separate(newname, into=c("name","qname"),":") %>%
  mutate(group=ifelse(cgi < .2, "low_cgi","high_cgi")) 
 # mutate(group=ifelse(uncertainty_cgi > .04, "uncertain_cgi",group))
 # mutate(gb_cluster=ifelse(gb_cluster==1, "low_gb","high_gb")) %>%
  #unite(group,cgi_cluster,gb_cluster, sep=":") 
ggplot(clustered_reads,aes(x=cgi,fill=group))+geom_histogram()

clustered_reads$quantile = combined$quantile

#clustered_reads <- clustered_reads %>%
 # filter(!grepl("uncertain",group))



write.table(clustered_reads, "/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/revision_analysis/promoter_clustering/ClusteredReads_chrX_proteinCoding.tsv",
            quote=F,
            row.names=F,
            sep="\t")

clustered_reads <- read_tsv("/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/revision_analysis/promoter_clustering/ClusteredReads_chrX_proteinCoding.tsv")

# ggplot(clustered_reads,aes(x=cgi,y=gb,color=as.factor(quantile)))+geom_point(alpha=.05)

stat.sum <- clustered_reads %>%
  ungroup() %>%
  group_by(group) %>%
  summarise(sum=n())

stat <- clustered_reads %>%
  group_by(name) %>%
  mutate(total=n()) %>%
  ungroup() %>%
  group_by(name,group,quantile,total) %>%
  summarise(sum=n()) %>%
  mutate(frac=sum/total) %>%
  distinct()


ggplot(stat, aes(x=as.factor(quantile),y=frac,color=as.factor(quantile)))+geom_violin(width=1)+geom_boxplot(width=.1,outlier.shape = NA)+facet_wrap(~group)+theme_bw()

known_escapers <- read_csv("/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/revision_analysis/promoter_clustering/chrX/XCI_2014_paper.csv", col_names = c('name', 'escape')) %>%
  filter(escape %in% c("escape from XCI in 27 tissues","subject to XCI in 27 tissues"))

#escape <- stat %>%
 # mutate(escape=ifelse(name %in% known_escapers$gene, "yes","no"))

escape <- merge(stat,known_escapers, by="name")



write.table(escape, "/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/revision_analysis/promoter_clustering/ClusteredReadsFraction_chrX_EscapevsXCI.tsv",
            quote=F,
            row.names=F,
            sep="\t")

stat.escape <- clustered_reads %>%
  merge(known_escapers, by="name") %>%
  group_by(group,escape) %>%
  summarise(sum=n()) %>%
  ungroup()%>%
  group_by(escape) %>%
  mutate(total=sum(sum)) %>%
  ungroup() %>%
  mutate(frac=sum/total)
```

```{r}
clustered_reads <- read_tsv("/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/revision_analysis/promoter_clustering/ClusteredReads_chrX_proteinCoding.tsv")

regs <- regs.all %>%
  mutate(start=ifelse(strand=='+', gene_start,gene_end), end=ifelse(strand=='+', gene_end,gene_start)) %>%
  merge(known_escapers, by="name")# %>%
  #mutate(escape=ifelse(name %in% known_escapers$gene, "yes","no")) %>%
#  filter(transcripts > 2)

clustered_reads.sep <- clustered_reads %>%
  merge(known_escapers, by="name") %>%
  filter(name %in% regs$name) %>%
  group_by(escape,name,group) %>%
  mutate(n=n()) %>%
 # filter(n>10) %>%
  dplyr::select(qname,group,escape,name) %>%
  dplyr::rename("readname"=qname)
  

all.reads <- merge(reads, clustered_reads.sep,by="readname")
all.read.group1 <- all.reads %>%
  filter(group == "low_cgi") %>%
  dplyr::select(-c(group))

all.read.group2 <- all.reads %>%
  filter(group == "high_cgi") %>%
  dplyr::select(-c(group))

group1.cg <- mbedByCall(all.read.group1) %>%
  drop_na(mcall)
group1.freq <- group1.cg %>%
  mutate(end=start) %>%
  group_by(chrom,start,end) %>%
  summarise(methylated_frequency=mean(mcall)) %>%
  GRanges()

group2.cg <- mbedByCall(all.read.group2) %>%
  drop_na(mcall)
group2.freq <- group2.cg %>%
  mutate(end=start) %>%
  group_by(chrom,start,end) %>%
  summarise(methylated_frequency=mean(mcall)) %>%
  GRanges()

saveRDS(group1.freq, "/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/revision_analysis/promoter_clustering/xci/LOW_CGI_MethylationFrequency.rds")

saveRDS(group2.freq, "/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/revision_analysis/promoter_clustering/xci/HIGH_CGI_MethylationFrequency.rds")



flankn=1000
bodylen <- 4
# add methylation data
l1_regions <- regs %>%
  mutate(start = start - flankn, end = end + flankn) %>%
  GRanges()



ovl <- findOverlaps(group1.freq, GRanges(l1_regions))
genes.ovl <- as.data.frame(regs)[subjectHits(ovl),] %>%
  dplyr::mutate(genewidth = end - start) %>%
  dplyr::rename(reg_start = start, reg_end = end) %>%
  mutate("Direction"=strand)

# calculate distance from start and end, depending on strand 
chm13.ovl.group1 <- as.data.frame(group1.freq[queryHits(ovl),]) %>%
  bind_cols(genes.ovl) %>%
  dplyr::rename(seqnames = 1) %>%
  mutate(dist = ifelse(Direction=="+", start - reg_start, reg_end-start)) %>%
  mutate(dist = ifelse(dist < 0, dist/flankn,
      ifelse(dist < genewidth,
        bodylen * dist / genewidth,
        bodylen + (dist - genewidth)/flankn)), 
    dist = round(dist,2)
    ) %>%
  mutate(group = "group1")




ovl <- findOverlaps(group2.freq, GRanges(l1_regions))
genes.ovl <- as.data.frame(regs)[subjectHits(ovl),] %>%
  dplyr::mutate(genewidth = end - start) %>%
  dplyr::rename(reg_start = start, reg_end = end) %>%
  mutate("Direction"=strand)

# calculate distance from start and end, depending on strand 
chm13.ovl.group2 <- as.data.frame(group2.freq[queryHits(ovl),]) %>%
  bind_cols(genes.ovl) %>%
  dplyr::rename(seqnames = 1) %>%
  mutate(dist = ifelse(Direction=="+", start - reg_start, reg_end-start)) %>%
  mutate(dist = ifelse(dist < 0, dist/flankn,
      ifelse(dist < genewidth,
        bodylen * dist / genewidth,
        bodylen + (dist - genewidth)/flankn)), 
    dist = round(dist,2)
    ) %>%
  mutate(group = "group2")

chm13.ovl <- rbind(chm13.ovl.group1,chm13.ovl.group2) 

write.table(chm13.ovl, "/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/revision_analysis/promoter_clustering/ClusteredMethylationFrequency_chrX_proteinCoding_withUncertainty.tsv",
            quote=F,
            row.names=F,
            sep="\t")

# bin data to make plot pretty
n_windows=5000
chm13.ovl$cut = cut(chm13.ovl$dist, breaks=n_windows)

chm13.ovl.labs <- chm13.ovl %>%
  group_by(cut,group,escape) %>%
  summarise(med = median(methylated_frequency), top = quantile(methylated_frequency, 0.75), bot = quantile(methylated_frequency, 0.25), n_genes = length(methylated_frequency)) %>%
  mutate(x_tmp = str_sub(cut, 2, -2)) %>% 
  separate(x_tmp, c("min", "max"), sep = ",") %>% 
  mutate_at(c("min", "max"), as.double) %>%
  ungroup() %>%
  group_by(group,escape) %>%
  arrange(min) %>%
  mutate(med_smooth = rollmean(med, 15, NA),top_smooth = rollmean(top, 10, NA),bot_smooth = rollmean(bot, 10, NA))

# plot -- include annotation for number of repeats in each group
p <- ggplot(chm13.ovl.labs,aes( x = min, y = med_smooth, color=group, fill=group), alpha=.5)+
  geom_line(aes(y=med_smooth), alpha=.5, size=1) +
  geom_vline(xintercept = 0) +
  geom_vline(xintercept = bodylen)+
  labs( x = "Genomic Position", y = "Aggregated Methylation Frequency") +
  theme_classic()+facet_wrap(~escape,ncol = 1)+ylim(0,1)

#+scale_x_continuous(breaks= c(-1,0,bodylen,bodylen + 1), labels = c(paste0("-",flankn/1e3,"kb"),"Start","End",paste0("+",flankn/1e3,"kb")))+theme(legend.position = 'none')+facet_wrap(~escape,ncol = 1)+ylim(0,1)
p


ggsave(
  paste0(figs, "/CHM13_chrX_metaplot.pdf"),
  plot = p,
  scale = 1,
  width = 12,
  height = 8,
)
```

```{r}

saveRDS(all.read.group1, "/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/revision_analysis/promoter_clustering/xci/LOW_CGI_Reads.rds")

saveRDS(all.read.group2, "/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/revision_analysis/promoter_clustering/xci/HIGH_CGI_Reads.rds")


group1_reads <- all.read.group1 %>%
  dplyr::select(c(readname,escape,name)) %>%
  mutate(group="group1")
group2_reads <- all.read.group2 %>%
  dplyr::select(c(readname,escape,name)) %>%
  mutate(group="group2")

stat.reads <- rbind(group1_reads,group2_reads) %>%
  group_by(escape,name,group) %>%
  filter(name %in% regs$name) %>%
  summarise(n=n()) %>%
  filter(n>10) %>%
  group_by(name) %>%
  mutate(total=sum(n)) %>%
  mutate(frac=n/total) %>%
  filter(group == "group2") %>%
  arrange(frac) %>%
  ungroup() %>%
  mutate(order=row_number())

p1 <- ggplot(stat.reads, aes(x=order,y=frac,color=escape))+geom_point(alpha=.5,size=5)+ylim(0,1)

ggsave(
  paste0(figs, "/MethylationFraction.pdf"),
  plot = p1,
  scale = 1,
  width = 8,
  height = 5,
)

p2 <- ggplot(stat.reads, aes(x=frac, fill=escape))+geom_density(alpha=.5, adjust=.9)+xlim(0,1)

ggsave(
  paste0(figs, "/MethylationDensity.pdf"),
  plot = p2,
  scale = 1,
  width = 8,
  height = 5,
)

stat.all <- rbind(group1_reads,group2_reads) %>%
  filter(name %in% regs$name) %>%
  group_by(group,escape) %>%
  summarise(per_group=n()) %>%
  ungroup() %>%
  group_by(escape) %>%
  mutate(total=sum(per_group)) %>%
  mutate(frac=per_group/total) %>%
  ungroup() 


    
```



```{r}
gene="AR"
acot9 <- clustered_reads %>%
  dplyr::filter(name==gene) %>%
  dplyr::select(c(group,qname))

cgcalls.sub <- as.data.frame(cgcalls) %>%
  filter(qname %in% acot9$qname)

cpg_runs <-getRuns(cgcalls.sub, maxGap = 150)
all_runs <- merge(cpg_runs,clustered_reads, by = "qname" )

cpg_runs.ordered <- order_reads_facet(all_runs)



cpg_runs_plot <- cpg_runs.ordered$x %>%
    mutate(m = ifelse(values == 1, "Methylated","Unmethylated")) %>%
    mutate(mod = "CpG")
  
pal <- pal_npg("nrc")(10)
meth_pal <- c(pal[8],pal[4]) 

coords <- as.data.frame(regs.all) %>%
  filter(name == gene)  %>%
  mutate(start=ifelse(strand=="+", cgi_start,cgi_end))%>%
  mutate(end=gene_end)
#coords$end[1]=3300000
g <- ggplot(cpg_runs_plot,aes(xmin = start, xmax = end, ymin = ymin, ymax = ymax))+ geom_rect(data = cpg_runs.ordered$bounds, fill = "grey80")+ 
    geom_rect(aes(fill = m)) +scale_fill_manual(name = "State", values = meth_pal)+theme_classic()+facet_wrap(~group,ncol=1)+theme(legend.position = "bottom")+ theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), panel.spacing = unit(2, "lines"))+coord_cartesian(xlim=c(coords$end[1], coords$start[1]))+geom_vline(xintercept = c(coords$cgi_start[1],coords$cgi_end[1]), linetype = "dotted")
g

figs="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/revision_analysis/promoter_clustering/chrX/"
base="chrX"
ggsave(paste0(figs, base,".", coords$name[1], ".pdf"),
       g,
       height = 8, 
       width=12)
```


```{r}

rep="SVAE"

l1s <- read_tsv("/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly/TE/revision/new_deeptools_beds/SVAE_HOTnot_CHM13ABpool_anti-sense_htmpSorted_3Prime.bed") %>%
  dplyr::rename("chr"=`#chrom`) %>%
  dplyr::rename("Direction"=strand) %>%
  mutate(len=abs(end-start))
gene_regs <- regs %>%
  mutate(start=ifelse(strand=="+", gene_start,gene_end)) %>%
  mutate(end=ifelse(strand=="+", gene_end,gene_start)) 
  

l1_genes <- FindOvls(GRanges(l1s), GRanges(gene_regs)) %>%
  select(-c(gene_start, gene_end)) %>%
  mutate(ID=row_number())

flankn <- 100
bodylen <- 4000

# overlap with m
# overlap with methylation calls and normalize by size 
l1_regions <- l1_genes %>%
  dplyr::mutate(gene_start = start, gene_end = end) %>%
  mutate(reg_end = ifelse(Direction == "+", end + flankn, start-flankn)) %>%
  mutate(reg_start = ifelse(Direction=="+", end - bodylen, start+bodylen)) %>%
  mutate(start=ifelse(Direction=="+", reg_start,reg_end)) %>%
  mutate(end=ifelse(Direction=="+", reg_end,reg_start)) %>%
  GRanges()

ovl <- findOverlaps(GRanges(group1.freq), l1_regions)
genes.ovl <- as.data.frame(l1_genes)[subjectHits(ovl),] %>%
  dplyr::mutate(genewidth = end - start) %>%
  dplyr::rename(gene_start = start, gene_end = end) 

chm13.ovl1 <- as.data.frame(GRanges(group1.freq)[queryHits(ovl),]) %>%
  bind_cols(genes.ovl) %>%
  dplyr::rename(seqnames = 1) %>%
  dplyr::mutate(dist = ifelse(Direction == "+",gene_end - start,start - gene_start),
                #dist = ifelse(dist < 0, dist/flankn,
                #              ifelse(dist < genewidth,
                #                     bodylen * dist / genewidth,
                #                     bodylen + (dist - genewidth)/flankn)), 
                dist = plyr::round_any(dist,5)*-1)
#  )


ovl <- findOverlaps(GRanges(group2.freq), l1_regions)
genes.ovl <- as.data.frame(l1_genes)[subjectHits(ovl),] %>%
  dplyr::mutate(genewidth = end - start) %>%
  dplyr::rename(gene_start = start, gene_end = end) 

chm13.ovl2 <- as.data.frame(GRanges(group2.freq)[queryHits(ovl),]) %>%
  bind_cols(genes.ovl) %>%
  dplyr::rename(seqnames = 1) %>%
  dplyr::mutate(dist = ifelse(Direction == "+",gene_end - start,start - gene_start),
                #dist = ifelse(dist < 0, dist/flankn,
                #              ifelse(dist < genewidth,
                #                     bodylen * dist / genewidth,
                #                     bodylen + (dist - genewidth)/flankn)), 
                dist = plyr::round_any(dist,10)*-1)
#  )

chm13.ovl1 <- chm13.ovl1 %>%
  mutate(group="group1")
chm13.ovl2 <- chm13.ovl2  %>%
  mutate(group="group2")
chm13.ovl <- rbind(chm13.ovl1,chm13.ovl2) 

n_windows=1000
chm13.ovl$cut = cut(chm13.ovl$dist, breaks=n_windows)

chm13.ovl.labs <- chm13.ovl %>%
  group_by(cut,group,escape) %>%
  summarise(med = median(methylated_frequency),ID=ID,escape=escape) %>%
  mutate(x_tmp = str_sub(cut, 2, -2)) %>% 
  separate(x_tmp, c("min", "max"), sep = ",") %>% 
  mutate_at(c("min", "max"), as.double) %>%
  ungroup() %>%
  distinct() %>%
  group_by(group,escape) %>%
  arrange(min) %>%
  mutate(med_smooth = rollmean(med, 2, NA))

# plot frequency with loess --- if doing SINEs don't use loess unless u want to die 
freqplot <- ggplot(chm13.ovl.labs,aes(x=min,y=med_smooth, color=group))+theme(legend.position = "left", legend.direction="vertical",axis.title.x=element_blank())+ylim(0,1)+ geom_smooth(method="loess",se=F, span=.3)+facet_wrap(~escape)#+geom_line()

# plot heatmap
plot <- ggplot(chm13.ovl,aes(x=dist,y=-ID,fill=methylated_frequency))+ geom_tile()+scale_fill_gradient(low = "blue", high = "red", na.value = NA, limits=c(0,1))+theme(legend.position = "left", legend.direction="vertical")+theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), axis.title.x=element_blank(),axis.title.y=element_blank())+facet_wrap(~group+escape,ncol=1)#+scale_x_continuous(breaks= c(-1,0,bodylen,bodylen + 1), labels = c(paste0("-",flankn/1e3,"kb"),"TTS","TES",paste0("+",flankn/1e3,"kb")))

# stack density and heatmap
plot_grid(freqplot, plot, ncol=1, align = "v", rel_heights=c(1/3,1))


ggsave(
  paste0(figs, "/", rep, "_Heatmap.pdf"),
  plot = last_plot(),
  scale = 1,
  width = 8,
  height = 12,
)

ggsave(
  paste0(figs, "/" , rep, "_Heatmap.png"),
  plot = last_plot(),
  scale = 1,
  width = 8,
  height = 12,
)
```

