---
title: "R Notebook"
output: html_notebook
---


```{r load_libs}
source("~/T2T-Epigenetics/utils/ilee_plot_utils.R")
source("~/T2T-Epigenetics/utils/methylation_R_utils.R")
library(tidyverse)
library(cowplot)
library(BSgenome.t2t.v1.0.release)
library(bsseq)
library(zoo)
```

We want to look at methylation differences between different types of TEs -- will start with the L1 LINEs 

Load the bed files savannah made with metadata about subtype and "HOT" or not
"HOT" refers to if the L1 is active 

```{r load_dat}
dat="/kyber/Data/Nanopore/Analysis/gmoney/CHM13/v1.0_final_assembly"
reps <- read_tsv(paste0(dat, "/TE/beds/L1-LINE_chm13v1_polish-033121_HOT-syn-CEN-subTELO-cenSAT-CBmotif-avgMeth_V3.bed"), col_names = c("chrom", "start", "end", "length", "query", "strand", "rep_type", "rep_class_rep_family", "position_begin", "in_end", "repeat_left", "SW", "perc_div", "perc_del", "perc_ins", "HOT", "no_syn", "CEN", "subtelo", "cenSAT", "CENPB_pJalpha", "average_methylation_frequency")) %>%
  mutate(direction=case_when(strand == "+" ~"+", 
                          strand == "C" ~ "-", 
                          TRUE ~ "*")) %>%
  mutate(name = ifelse(grepl("L1M", rep_type), "L1M", rep_type)) %>%
  mutate(name = ifelse(grepl("L1P", name), "L1P", name)) %>%
  mutate(name = ifelse(grepl("L1H", name), "L1H", name)) %>%
  select(-c(strand)) %>%
  filter(name %in% c("L1M", "L1P", "L1H")) %>%
  mutate(centro=case_when(CEN == "." ~ "no", 
                          TRUE ~ "yes")) %>%
  mutate(ID=row_number())

chm13_meth <- read_tsv(paste0(dat, "/methylation_calls/methylation_frequency_50kb_split.tsv")) %>%
  mutate(called_sites_unmethylated = called_sites - called_sites_methylated)

```




```{r}
flankn <- 1000
bodylen <- 6

l1_regions <- reps %>%
  mutate(start = start - flankn, end = end + flankn) %>%
  GRanges()

ovl <- findOverlaps(GRanges(chm13_meth), l1_regions)
genes.ovl <- as.data.frame(reps)[subjectHits(ovl),] %>%
  dplyr::mutate(genewidth = end - start) %>%
  dplyr::rename(gene_start = start, gene_end = end) 

chm13.ovl <- as.data.frame(GRanges(chm13_meth)[queryHits(ovl),]) %>%
  bind_cols(genes.ovl) %>%
  dplyr::rename(seqnames = 1) %>%
  dplyr::mutate(dist = ifelse(direction == "+",start - gene_start, gene_end - start),
    dist = ifelse(dist < 0, dist/flankn,
      ifelse(dist < genewidth,
        bodylen * dist / genewidth,
        bodylen + (dist - genewidth)/flankn)), 
    dist = round(dist,3)
    )
```

```{r}
library(ggpubr)

t2t.cpg.loci <- findLoci(pattern = "CG",
                         subject = BSgenome.t2t.v1.0.release::BSgenome.t2t.v1.0.release, 
                         strand = "+")
length(t2t.cpg.loci)

ovl <- FindOvls(t2t.cpg.loci, GRanges(reps)) %>%
  mutate(average_methylation_frequency=as.numeric(average_methylation_frequency)) %>%
  group_by(ID) %>%
  summarise(num_cpg=n()/length, name=name,average_methylation_frequency=average_methylation_frequency) %>%
  distinct()

my_comparisons <- list( c("L1H", "L1M"), c("L1H", "L1P"), c("L1M", "L1P") )
ggplot(ovl, aes(x=name, y=num_cpg, fill=name))+geom_boxplot()+ stat_compare_means(comparisons = my_comparisons)

ggsave(
  paste0(dat, "/figures/evol_meth/TE/","L1_HvsMvsP_CpG_KWtestV3.pdf"),
  plot = last_plot(),
  scale = 1,
  width = 5,
  height = 5,
)

ggplot(ovl, aes(x=name, y=average_methylation_frequency, fill=name))+geom_boxplot()+ stat_compare_means(comparisons = my_comparisons)

ggsave(
  paste0(dat, "/figures/evol_meth/TE/","L1_HvsMvsP_averagemethylation_KWtestV3.pdf"),
  plot = last_plot(),
  scale = 1,
  width = 5,
  height = 5,
)
```

```{r}

n_windows=500
chm13.ovl$cut = cut(chm13.ovl$dist, breaks=n_windows)

chm13.ovl.labs <- chm13.ovl %>%
  filter(name=="L1H") %>%
  group_by(cut,HOT) %>%
 # mutate(methylated_frequency = num_meth/(num_meth+num_unmeth)) %>%
  summarise(med = median(methylated_frequency), top = quantile(methylated_frequency, 0.75), bot = quantile(methylated_frequency, 0.25), n_genes = length(methylated_frequency)) %>%
  mutate(x_tmp = str_sub(cut, 2, -2)) %>% 
  separate(x_tmp, c("min", "max"), sep = ",") %>% 
  mutate_at(c("min", "max"), as.double) %>%
  ungroup() %>%
  group_by(HOT) %>%
  arrange(min) %>%
  mutate(med_smooth = rollmean(med, 10, NA))

p <- ggplot(chm13.ovl.labs,aes( x = min, y = med_smooth, color=HOT, fill=HOT), alpha=.5)+
#  geom_smooth(method = "loess", span = 0.12,se = F)+
#  geom_smooth(se=T)+lims( y = c(0,1)) +
  geom_ribbon(aes(ymin=bot, ymax=top), alpha=0.2, size=0)+
  geom_point()+
  geom_line(aes(y=med_smooth), alpha=.5, size=1) +
  geom_vline(xintercept = 0) +
  geom_vline(xintercept = bodylen)+
  labs( x = "Genomic Position", y = "Aggregated Methylation Frequency") +
theme(legend.background = element_rect(color = "black")) +
  theme_classic()+scale_x_continuous(breaks= c(-1,0,bodylen,bodylen + 1), labels = c(paste0("-",flankn/1e3,"kb"),"Start","End",paste0("+",flankn/1e3,"kb")))
p


ggsave(
  paste0(dat, "/figures/evol_meth/TE/","L1H_hot_not_1.25kbV3.pdf"),
  plot = p,
  scale = 1,
  width = 15,
  height = 10,
)

```



```{r}

reps_sub <- reps %>%
  filter(name=="L1H") %>%
  mutate(censat_name = ifelse(grepl("ct", cenSAT), ".", cenSAT)) %>%
  mutate(censat_name = ifelse(grepl("dhor", censat_name), "DHOR", censat_name)) %>%
  mutate(censat_name = ifelse(grepl("hor", censat_name), "HOR", censat_name)) %>%
  mutate(censat_name = ifelse(grepl("mon", censat_name), "MON", censat_name)) %>%
  mutate(censat_name = ifelse(grepl("hsat", censat_name), "HSat", censat_name)) %>%
  mutate(censat_name = ifelse(grepl("bsat", censat_name), "BSat", censat_name))
library(ggpubr)
ovl <- FindOvls(GRanges(chm13_meth), GRanges(reps_sub)) %>%
  mutate(average_methylation_frequency=as.numeric(average_methylation_frequency)) %>%
  group_by(ID,censat_name,HOT) %>%
 # summarise(num_cpg=n()/length, name=name,average_methylation_frequency=average_methylation_frequency) %>%
  distinct()

#my_comparisons <- list( c(".", "DHOR"), c(".", "HOR"), c(".", "MON") )
p <- ggplot(ovl, aes(x=name, y=methylated_frequency, fill=censat_name))+geom_violin(position = position_dodge())+geom_boxplot(width=.1,position = position_dodge(width =0.9))+facet_wrap(~HOT)

ggsave(
  paste0(dat, "/figures/evol_meth/TE/L1_hotvsnot_violinV3.pdf"),
  plot = p,
  scale = 1,
  width = 10,
  height = 5,
)

```

```{r}

  
SAT = c( "DHOR", "HOR", "MON")
censat = read_tsv(paste0(dat, "/annotations/t2t_cenAnnotation.v2.021621FORMATTED.bed"), col_names = F) %>%
    mutate(name = ifelse(grepl("gap", X4), "gap-rDNA", X4)) %>%
    filter(name %in% SAT) %>%
    dplyr::rename("chr" =1, "start" = 2 ,"end" =3) %>%
  GRanges()
  
library(ggpubr)
ovl2 <- FindOvls(GRanges(chm13_meth), censat) %>%
  select(name, methylated_frequency) %>% 
  mutate(HOT="All")

keepi <- findOverlaps(GRanges(chm13_meth),censat)
freq.matched <- GRanges(chm13_meth)[-queryHits(keepi)]


nonovl <- as.data.frame(freq.matched) %>%
  mutate(name=".") %>%
  select(name, methylated_frequency)%>% 
  mutate(HOT="All")


ovl <- ovl %>%
  dplyr::mutate(name=censat_name) %>%
  filter(name != "BSat") %>%
  filter(name != "HSat") %>%
  select(c(name, methylated_frequency,HOT)) 

allovls <- rbind(ovl, ovl2,nonovl)

my_comparisons <- list( c(".", "DHOR"), c(".", "HOR"), c(".", "MON") )
q <- ggplot(allovls, aes(x=HOT, y=methylated_frequency, fill=HOT))+geom_violin(position = position_dodge())+geom_boxplot(width=.1,position = position_dodge(width =0.9))+facet_wrap(~name)

ggsave(
  paste0(dat, "/figures/evol_meth/TE/","L1_region_hotvsnotV3.pdf"),
  plot = q,
  scale = 1,
  width = 10,
  height = 10,
)

```

```{r}

reg="MON"
hor_lines <- reps_sub %>%
 # filter(ID %in% IDs) %>%
  filter(censat_name==reg)

write.table(hor_lines, file=paste0(dat, "/figures/evol_meth/TE/", reg, "_hot_not_metadata.tsv"), quote=F, sep='\t', row.names=F, col.names=T)


overlaps <- FindOvls(GRanges(chm13_meth), GRanges(hor_lines))


ggplot(overlaps, aes(x=start, y=methylated_frequency,color=seqnames))+geom_point()+facet_wrap(~ID,scales="free",ncol=1)+geom_smooth(span=.4,method = "loess",se=F)+ylim(0,1)

ggsave(
  paste0(dat, "/figures/evol_meth/TE/", reg, "_l1_hotvsnotALLV3.pdf"),
  plot = last_plot(),
  scale = 1,
  width = 10,
  height = 15,
)
```


```{r}
n_windows=500
chm13.ovl$cut = cut(chm13.ovl$dist, breaks=n_windows)

chm13.ovl.labs <- chm13.ovl %>%
  filter(name=="L1H") %>%
  mutate(censat_name = ifelse(grepl("ct", cenSAT), ".", cenSAT)) %>%
  mutate(censat_name = ifelse(grepl("dhor", censat_name), "DHOR", censat_name)) %>%
  mutate(censat_name = ifelse(grepl("hor", censat_name), "HOR", censat_name)) %>%
  mutate(censat_name = ifelse(grepl("mon", censat_name), "MON", censat_name)) %>%
  mutate(censat_name = ifelse(grepl("hsat", censat_name), "HSat", censat_name)) %>%
  mutate(censat_name = ifelse(grepl("bsat", censat_name), "BSat", censat_name))  %>%
  filter(censat_name != "BSat") %>%
  filter(censat_name != "HSat") %>%
  group_by(cut,HOT,censat_name) %>%
 # mutate(methylated_frequency = num_meth/(num_meth+num_unmeth)) %>%
  summarise(med = median(methylated_frequency), top = quantile(methylated_frequency, 0.75), bot = quantile(methylated_frequency, 0.25), n_genes = length(methylated_frequency)) %>%
  mutate(x_tmp = str_sub(cut, 2, -2)) %>% 
  separate(x_tmp, c("min", "max"), sep = ",") %>% 
  mutate_at(c("min", "max"), as.double) %>%
  ungroup() %>%
  group_by(HOT,censat_name) %>%
  arrange(min) %>%
  mutate(med_smooth = rollmean(med, 10, NA))

p <- ggplot(chm13.ovl.labs,aes( x = min, y = med_smooth, color=HOT, fill=HOT), alpha=.5)+
#  geom_smooth(method = "loess", span = 0.12,se = F)+
#  geom_smooth(se=T)+lims( y = c(0,1)) +
  geom_ribbon(aes(ymin=bot, ymax=top), alpha=0.2, size=0)+
  geom_point()+
  geom_line(aes(y=med_smooth), alpha=.5, size=1) +
  geom_vline(xintercept = 0) +
  geom_vline(xintercept = bodylen)+
  labs( x = "Genomic Position", y = "Aggregated Methylation Frequency") +
theme(legend.background = element_rect(color = "black")) +
  theme_classic()+scale_x_continuous(breaks= c(-1,0,bodylen,bodylen + 1), labels = c(paste0("-",flankn/1e3,"kb"),"Start","End",paste0("+",flankn/1e3,"kb")))+facet_wrap(~censat_name)
p


ggsave(
  paste0(dat, "/figures/evol_meth/TE/","L1H_hot_not__regions_1kbV3.pdf"),
  plot = p,
  scale = 1,
  width = 15,
  height = 10,
)

```

